cpp/src_prims/selection/knn.h

```diff
@@ -18,179 +18,598 @@
+ * @param n_labels number of unique labels
+ * @param target_val the label value to search for in unique_labels
+ */
+template <typename IdxType = int>
+__device__ int label_binary_search(IdxType *unique_labels, IdxType n_labels,
+                                   IdxType target_val) {
+  int out_label_idx = -1;
+
+  int level = 1;
+  int cur_break_idx = round(n_labels / (2.0 * level));
+  while (out_label_idx == -1) {
+    int cur_cached_label = unique_labels[cur_break_idx];
+
+    // If we found our label, terminate
+    if (cur_cached_label == target_val) {
+      return cur_break_idx;
+
+      // check left neighbor
+    } else if (cur_break_idx > 0 &&
+               unique_labels[cur_break_idx - 1] == target_val) {
+      return cur_break_idx - 1;
+
+      // check right neighbor
+    } else if (cur_break_idx < n_labels - 1 &&
+               unique_labels[cur_break_idx + 1] == target_val) {
+      return cur_break_idx + 1;
+
+      // traverse
+    } else {
+      level += 1;
+
+      int subtree = round(n_labels / (2.0 * level));
+      if (target_val < cur_cached_label) {
+        // take left subtree
+        cur_break_idx -= subtree;
       } else {
-        std::stringstream ss;
-        ss << "Input memory for " << ptr
-           << " failed. isDevice?=" << att.devicePointer << ", N=" << sizes[i];
-        std::cout << "Exception: " << ss.str() << std::endl;
+        // take right subtree
+        cur_break_idx += subtree;
       }
     }
   }
+  return -1;
+}
 
-  merge_tables<faiss::CMin<float, IntType>>(
-    long(n), k, n_params, result_D, result_I, all_D, all_I, id_ranges->data());
+template <typename OutType = float>
+__global__ void class_probs_kernel(OutType *out, const int64_t *knn_indices,
+                                   const int *labels, int *unique_labels,
+                                   int n_uniq_labels, size_t n_samples,
+                                   int n_neighbors) {
+  int row = (blockIdx.x * blockDim.x) + threadIdx.x;
+  int i = row * n_neighbors;
 
-  MLCommon::updateDevice(res_D, result_D, k * size_t(n), s);
-  MLCommon::updateDevice(res_I, result_I, k * size_t(n), s);
+  extern __shared__ int label_cache[];
 
-  delete all_D;
-  delete all_I;
+  if (threadIdx.x == 0) {
+    for (int j = 0; j < n_uniq_labels; j++) label_cache[j] = unique_labels[j];
+  }
 
-  delete result_D;
-  delete result_I;
-};
+  __syncthreads();
+
+  if (row >= n_samples) return;
+
+  for (int j = 0; j < n_neighbors; j++) {
+    int64_t neighbor_idx = knn_indices[i + j];
+    int out_label = labels[neighbor_idx];
+
+    // Trading off warp divergence in the outputs so that we don't
+    // need to copy / modify the label memory to do these mappings.
+    // Found a middle-ground between between using shared memory
+    // for the mappings.
+    int out_label_idx =
+      label_binary_search(label_cache, n_uniq_labels, out_label);
+
+    int out_idx = row * n_uniq_labels + out_label_idx;
+    out[out_idx] += 1.0f;
+  }
+}
```

Comment:
instead of writing separate `class_probs_kernel` and `class_vote_kernel`, why not also add the logic of finding the best probability here itself?

Reply:
I could certainly do the probabilities and the voting in the same kernel. I would still need a kernel that can just do the prediction of the probabilities, though, since the Kneighbors classifier requires a `predict_proba()` function.

Reply:
I have created issue #1397 to address this.